### DISCUSSION TITLE
Provide encoding for scripts

### DISCUSSION THREAD

**Discussion by someth2say** [jbangdev/jbang#1399] [Q&A] [answered]
Hi there!

I have a situation here, and I could use some help to find the best approach.
I have a JBang script that outputs using unicode emoji, for example:
```
@Command(name = "sk", mixinStandardHelpOptions = true, ...,  footer = "Made with 💘 by GLS Learning Content Services", ... )
```
Those unicode emojis are `UTF8` encoded. 
But now, I have a user whose default encoding is `US-ASCII`, and things started going bananas.
```
@Command(name = "sk", mixinStandardHelpOptions = true, ...,  footer = "Made with ???? by GLS Learning Content Services", ... )
                                                                                 ^
/....java:72: error: unmappable character (0x9F) for encoding US-ASCII
```

In plain java, there are two ways for providing the right encoding:
1) Use `java -Dfile.encoding=UTF8`
I assumed that I can pass it to JBang via `JBANG_JAVA_OPTIONS="-Dfile.encoding=UTF8"`, but didn´t work. 
AFAICS, it seems that JBang does not honor those java options when compiling the scripts (despite I failed to find the exact place for that).

2) Use ` JAVA_TOOL_OPTIONS="-Dfile.encoding=UTF8"` environment variable.
This affects the whole JVM, and indeed works fine.
But, as per JVM definition, this generates stderr messages:
```
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
```
Unluckily, I cannot suppress nor ignore the messages, hence the script output gets polluted.

My question is: Is there a way to instruct JBang to use the proper encoding when loading script sources?

**Comment by quintesse** [comment, marked as answer]
Have you tried `jbang -D=file.encoding=UTF8`? Not sure if it works but it's worth a try.

**Reply by someth2say** [reply]
No luck.
`jbang  -D=file.encoding=UTF8 run ...` spits an "unknown option".
`jbang run -D=file.encoding=UTF8 ...` does not barf, but it seems to ignore the `-D`.

**Reply by quintesse** [reply]
Can you try adding the line `//JAVAC_OPTIONS -encoding=UTF8` to the source file @someth2say ?

**Reply by someth2say** [reply]
Almost!
With `//JAVAC_OPTIONS -encoding=UTF8` I got a compiler error, but  `//JAVAC_OPTIONS -encoding UTF8` works!
That solves my problem!

Thanks, @quintesse !

**Reply by quintesse** [reply]
Oop, yes, that was a typo on my part :-)
Ok, glad that works at least. Which means this issue would be useful too: https://github.com/jbangdev/jbang/issues/1397

**Comment by maxandersen** [comment]
-D should work. 

What does --verbose print ?

**Reply by quintesse** [reply]
@maxandersen I see that the option is not being passed to `javac`. Didn't we use to do that? I'm not sure if that's a regression or that we decided at some point to remove it or if I'm remembering wrong and we never did that?

**Reply by quintesse** [reply]
Ah, `javac` doesn't have a generic `-D` option 😅

**Reply by maxandersen** [reply]
Sounds like regression to me but not 100% sure.

**Reply by maxandersen** [reply]
//JAVAC_OPTIONS -Dfile... 

Should work.

**Reply by quintesse** [reply]
For javac it's actually `-encoding`

**Reply by quintesse** [reply]
> Sounds like regression to me but not 100% sure.

Yeah, I thought so too for a moment, but `javac` actually only supports 2 very specific `-D` options. If you try to do something like `-Dfoo` it throws an error.

Edit: although you can do `javac -J-Dfoo=bar`, but I don't think we ever did that. It seems somewhat tricky TBH to be passing the same -D options to both the compiler and the runtime.
